SerializeLoc: serialize basic decl source location information to .swiftsourceinfo file

nkcsgexi opened the pull request:
After setting up the .swiftsourceinfo file, this patch starts to actually serialize and de-serialize source locations for declaration. The binary format of .swiftsourceinfo currently contains these three records:

BasicDeclLocs: a hash table mapping from a USR ID to a list of basic source locations. The USR id could be retrieved from the following DeclUSRs record using an actual decl USR. The basic source locations include a file ID and the results from Decl::getLoc(), ValueDecl::getNameLoc(), Decl::getStartLoc() and Decl::getEndLoc(). The file ID could be used to retrieve the actual file name from the following SourceFilePaths record. Each location is encoded as a line:column pair.

DeclUSRS: a hash table mapping from USR to a USR ID used by location records.

SourceFilePaths: a hash table mapping from a file ID to actual file name.

BasicDeclLocs should be sufficient for most diagnostic cases. If additional source locations are needed, we could always add new source location records without breaking the backward compatibility. When de-serializing the source location from a module-imported decl, we calculate its USR, retrieve the USR ID from the DeclUSRS record, and use the USR ID to look up the basic location list in the BasicDeclLocs record.

For more details about .swiftsourceinfo file: https://forums.swift.org/t/proposal-emitting-source-information-file-during-compilation

rdar://56167685

nkcsgexi committed:
SerializeLoc: serialize basic decl source location information to .swiftsourceinfo file

After setting up the .swiftsourceinfo file, this patch starts to actually serialize
and de-serialize source locations for declaration. The binary format of .swiftsourceinfo
currently contains these three records:

BasicDeclLocs: a hash table mapping from a USR ID to a list of basic source locations. The USR id
could be retrieved from the following DeclUSRs record using an actual decl USR. The basic source locations
include a file ID and the results from Decl::getLoc(), ValueDecl::getNameLoc(), Decl::getStartLoc() and Decl::getEndLoc().
The file ID could be used to retrieve the actual file name from the following SourceFilePaths record.
Each location is encoded as a line:column pair.

DeclUSRS: a hash table mapping from USR to a USR ID used by location records.

SourceFilePaths: a hash table mapping from a file ID to actual file name.

BasicDeclLocs should be sufficient for most diagnostic cases. If additional source locations
are needed, we could always add new source location records without breaking the backward compatibility.
When de-serializing the source location from a module-imported decl, we calculate its USR, retrieve the USR ID
from the DeclUSRS record, and use the USR ID to look up the basic location list in the BasicDeclLocs record.

For more details about .swiftsourceinfo file: https://forums.swift.org/t/proposal-emitting-source-information-file-during-compilation

nkcsgexi commented on test/Serialization/comments.swift:
```diff
@@ -58,14 +60,32 @@ public protocol P1 { }
+// SECOND: Inputs/def_comments.swift:28:13: Var/decl_var_2 RawComment=none BriefComment=none DocCommentAsXML=none
+// SECOND: Inputs/def_comments.swift:28:25: Var/decl_var_3 RawComment=none BriefComment=none DocCommentAsXML=none
+// SECOND: Inputs/def_comments.swift:28:25: Var/decl_var_3 RawComment=none BriefComment=none DocCommentAsXML=none
+// SECOND: NonExistingSource.swift:100000:13: Func/functionAfterPoundSourceLoc
```
@rintaro as you have suggested, a test case for `#sourceLocation` is added here.

nkcsgexi committed:
test: add a test for serializing decl source locations under batch mode

nkcsgexi committed:
test: add a test for source location serialization for wmo mode

nkcsgexi committed:
SerializeLoc: ensure we can load serialized source locations when we are importing a module from a Swift interface file

nkcsgexi committed:
SerializeLoc: add some comments about record structures. NFC

nkcsgexi committed:
SerializeLoc: include source locations for double underscored symbols in .swiftsourceinfo file

For .swiftdoc file, we don't expose doc-comments for underscored symbols. But this
seems to be an unnecessary constraint on .swiftsourceinfo file since we put
these symbols in .swiftinterface files anyway.

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci Please test compiler performance

nkcsgexi commented:
@swift-ci Please Build Toolchain macOS Platform

jrose-apple commented on include/swift/AST/RawComment.h:
```diff
@@ -78,6 +78,19 @@ struct CommentInfo {
+struct BasicDeclLocs {
+  StringRef SourceFilePath;
+  Optional<LineColumn> Loc;
+  Optional<LineColumn> NameLoc;
```
I don't think we have any decls where Loc and NameLoc are different. Is this just to account for decls without names?

nkcsgexi replied:
We have separate APIs `getLoc()` and `getNameLoc()` for `ValueDecl`, so they could potentially be different. IMO, we shouldn't hardcode that they are usually identical to the format.

jrose-apple replied:
I'll flip it around: there are zero cases where `getLocFromSource()` does not return `getNameLoc()` if there is a `getNameLoc()`, and we could put that in the AST verifier.

jrose-apple commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -2202,34 +2463,48 @@ Optional<CommentInfo> ModuleFile::getCommentForDecl(const Decl *D) const {
+  return *I;
+}
+
+uint32_t ModuleFile::getDeclUSRId(StringRef USR) const {
```
Hmm, I guess this is future-proofing for other tables keyed by USR, so that the string doesn't get repeated in the keys of every table? It does seems like if you have an integer ID you don't need a hash table any more, though.

jrose-apple commented on lib/Serialization/ModuleFormat.h:
```diff
@@ -638,6 +638,19 @@ enum BlockID {
+  /// source location blocks.
+  ///
+  /// This is part of a stable format and must not be renumbered!
+  MODULE_SOURCEINFO_BLOCK_ID = 192,
```
I thought you decided these *weren't* going to be considered a stable format, since they aren't installed? As such I'm not sure you need to put them after the moduledoc numbers.

jrose-apple replied:
If they are a stable format I'm a little more concerned about the format of these hash tables.

jrose-apple replied:
We talked about this in person. The plan is to attempt stability, but the consequences aren't so bad if we fail.

jrose-apple commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -263,10 +263,43 @@ std::error_code SerializedModuleLoaderBase::openModuleDocFile(
 }
 
+std::error_code
+SerializedModuleLoaderBase::openModuleSourceInfoFile(AccessPathElem ModuleID,
```
This seems like duplicated code; can you factor out the common parts instead?

nkcsgexi replied:
hmm, it seems we have similar duplications for other two module inputs too. Do you have any ideas where we should move it to?

jrose-apple replied:
I mean, here is fine, just generalize it.

jrose-apple commented on lib/Serialization/SourceInfoFormat.h:
```diff
@@ -0,0 +1,95 @@
+//===---= SourceInfoFormat.h - The internals of swiftsourceinfo files ----===//
```
Missing the `-*- c++ -*-`.

jrose-apple commented on lib/Serialization/SourceInfoFormat.h:
```diff
@@ -0,0 +1,95 @@
+/// to make backwards-compatible changes using the LLVM bitcode format.
+///
+/// \sa DECL_LOCS_BLOCK_ID
+namespace sourceinfo_block {
```
Is it the DECL_LOCS block or the SOURCEINFO block?

jrose-apple commented on lib/Serialization/SourceInfoFormat.h:
```diff
@@ -0,0 +1,95 @@
+namespace sourceinfo_block {
+  enum RecordKind {
+    BASIC_DECL_LOCS = 1,
+    DECL_USRS = 96,
```
Why the jump to 96?

jrose-apple commented on lib/Serialization/SerializeDoc.cpp:
```diff
@@ -502,7 +502,390 @@ void serialization::writeDocToStream(raw_ostream &os, ModuleOrSourceFile DC,
   S.writeToStream(os);
 }
+namespace {
+struct LineColoumn {
```
Typo: Coloumn

jrose-apple commented on lib/Serialization/SerializeDoc.cpp:
```diff
@@ -502,7 +502,390 @@ void serialization::writeDocToStream(raw_ostream &os, ModuleOrSourceFile DC,
+    // EndLoc
+    dataLength += LineColumnLen;
+    endian::Writer writer(out, little);
+    writer.write<uint32_t>(keyLength);
```
If the length is constant you actually don't have to serialize it at all; returning it is sufficient.

jrose-apple commented on lib/Serialization/SerializeDoc.cpp:
```diff
@@ -502,7 +502,390 @@ void serialization::writeDocToStream(raw_ostream &os, ModuleOrSourceFile DC,
+    WRITE_LINE_COLUMN(Loc)
+    WRITE_LINE_COLUMN(NameLoc);
+    WRITE_LINE_COLUMN(StartLoc);
+    WRITE_LINE_COLUMN(EndLoc);
```
Inconsistent about the semicolons here. (I liked the lambda better, honestly.)

jrose-apple reviewed (commented):
A storm of comments on just the first commit, sorry.

davidungar reviewed (approved):
I just looked over it quickly, but nothing jumped out at me to comment upon. (I'll miss @jrose-apple 's reviews!) Let me know if you want another pair of eyes to go over it in depth.

nkcsgexi committed:
SerilizeLoc: refactor table structures per Jordan's comments

After this change, we only use one single hash table for USR to USR id
mapping. The basic source locations are an array of fixed length
records that could be retrieved by using the USR id since each
USR id is guaranteed to be associated with one basic location entry.

The source file paths are refactored to a blob of 0-terminated strings.
Decl locations use offset in this blob to refer to the source file path
where the decl was defined.

nkcsgexi commented:
@jrose-apple The most recent commit refactored the data structure as we discussed to use only one hash-table. Could you take another look?

nkcsgexi committed:
SerializeLoc: remove NameLoc entry in the serialized format

Having both Loc and NameLoc fields seems to be redundant.

nkcsgexi committed:
Frontend: change return type of CompilerInstance::getInputBuffersIfPresent to be Optional. NFC

nkcsgexi committed:
AST: rename printDeclUSRForModuleDoc to printDeclUSR. NFC

nkcsgexi committed:
Driver: rename Name.swiftmodule/Private directory to Name.swiftmodule/Project

This directory should be excluded during installation since the content is only
used for local development. swiftsourceinfo file is currently emitted to this directory.

nkcsgexi committed:
SerializeLoc: add an assertion to ensure we can use USR id as an index to access basic location records

nkcsgexi committed:
SerializeLoc: use decl_locs_block consistently. NFC

nkcsgexi commented:
@swift-ci please smoke test

brentdax commented on include/swift/Serialization/SerializedModuleLoader.h:
```diff
@@ -70,20 +71,30 @@ class SerializedModuleLoaderBase : public ModuleLoader {
       std::unique_ptr<llvm::MemoryBuffer> *ModuleBuffer,
-      std::unique_ptr<llvm::MemoryBuffer> *ModuleDocBuffer) = 0;
+      std::unique_ptr<llvm::MemoryBuffer> *ModuleDocBuffer,
+      std::unique_ptr<llvm::MemoryBuffer> *ModuleSourceInfoBuffer) = 0;
```
These now-three files are passed together so often that I'm beginning to think we should have an abstraction to represent this grouping. `ModuleFiles<StringRef> ModuleFilenames`, `ModuleFiles<std::unique_ptr<llvm::MemoryBuffer> *> ModuleBuffers`.

brentdax commented on include/swift/AST/RawComment.h:
```diff
@@ -79,16 +79,17 @@ struct CommentInfo {
-  uint32_t Column;
+  uint32_t Line = 0;
+  uint32_t Column = 0;
+  bool isValid() const { return Line && Column; }
```
Would it make sense to assert somewhere that we never create a `LineColumn` with a zero `Line` but a nonzero `Column`, or vice versa? I worry about one of these fields ending up with garbage in it without us noticing because the other is zeroed, and then something reading that garbage and doing the wrong thing.

nkcsgexi replied:
hmm, not sure it worths it. We plan to use `LineColumn` here as an internal implementation detail just like `CommentInfo` defined above so users won't directly interact with this type. I'll open future PRs to hook `LineColumn` up with AST APIs like `getLoc()`, etc.

beccadax reviewed (approved):
I'm not going to pretend I had the attention span to read every line in detail, but it looked alright in my skim. If there's something specific you want me to examine more minutely, please let me know.

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi committed:
SerializeLoc: asserts USR id and basic decl location are always in sync

jrose-apple commented on lib/Frontend/Frontend.cpp:
```diff
@@ -509,17 +506,34 @@ CompilerInstance::getInputBufferAndModuleDocBufferIfPresent(
+  return ModuleBuffers(
+    std::move(*inputFileOrErr),
+    openModuleDoc(input).getValueOr(nullptr),
+    openModuleSourceInfo(input).getValueOr(nullptr)
```
Ooh, nice cleanup! Even better than what I had. :-)

jrose-apple commented on lib/Frontend/Frontend.cpp:
```diff
@@ -509,17 +506,34 @@ CompilerInstance::getInputBufferAndModuleDocBufferIfPresent(
+  llvm::SmallString<128> moduleSourceInfoFilePath(input.file());
+  llvm::sys::path::replace_extension(moduleSourceInfoFilePath,
+                                     file_types::getExtension(file_types::TY_SwiftSourceInfoFile));
+  std::string NonPrivatePath = moduleSourceInfoFilePath.str().str();
```
Nitpicks: this should be a SmallString too, and it should have a lowerCamelCase name, and I don't love that the name is described in terms of the *other* path.

jrose-apple replied:
Actually, I don't think we need to do the "Project" thing for the partial modules.

jrose-apple commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -1222,13 +1223,164 @@ bool ModuleFile::readModuleDocIfPresent() {
 }
 
+class ModuleFile::DeclUSRTableInfo {
+  public:
```
Nitpick: indentation

jrose-apple commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -1222,13 +1223,164 @@ bool ModuleFile::readModuleDocIfPresent() {
+    return llvm::djbHash(key, SWIFTSOURCEINFO_HASH_SEED);
+  }
+
+  static bool EqualKey(internal_key_type lhs, internal_key_type rhs) { return lhs == rhs; }
```
Nitpick: 80 cols

jrose-apple commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -1222,13 +1223,164 @@ bool ModuleFile::readModuleDocIfPresent() {
+
+  data_type ReadData(internal_key_type key, const uint8_t *data, unsigned length) {
+    assert(length == 4);
+    return *reinterpret_cast<const uint32_t*>(data);
```
\*sigh\* Technically this kind of reinterpret_cast is illegal in C++, please use the `endian::readNext` API anyway. (Also, it's important if we ever cross platform boundaries with endianness! For some reason!)

jrose-apple commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -1222,13 +1223,164 @@ bool ModuleFile::readModuleDocIfPresent() {
+  auto base = reinterpret_cast<const uint8_t *>(blobData.data());
+  return std::unique_ptr<SerializedDeclUSRTable>(
+    SerializedDeclUSRTable::Create(base + tableOffset, base + sizeof(uint32_t), base,
+                                   DeclUSRTableInfo()));
```
Nitpick: You can omit this last argument since DeclUSRTableInfo has a default constructor.

jrose-apple commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -1465,7 +1617,8 @@ ModuleFile::ModuleFile(
   }
-
+  // Read source info file.
+  readModuleSourceInfoIfPresent();
```
No error handling?

nkcsgexi replied:
Failing to load `.swiftsourceinfo` isn't a very big deal and probably is the default situation when loading a shipped Swift module. So I don't think it worths error handling.

jrose-apple commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -2202,34 +2355,63 @@ Optional<CommentInfo> ModuleFile::getCommentForDecl(const Decl *D) const {
+  assert(RecordOffset < BasicDeclLocsData.size());
+  assert(BasicDeclLocsData.size() % RecordSize == 0);
+  BasicDeclLocs Result;
+  auto *Record = reinterpret_cast<const uint32_t*>(BasicDeclLocsData.data() + RecordOffset);
```
Same deal, use explicit `endian::readNext` operations, please.

jrose-apple commented on lib/Serialization/ModuleFile.h:
```diff
@@ -540,6 +547,29 @@ class ModuleFile
+  bool readModuleSourceInfoIfPresent();
+
+  /// Read an on-disk decl hash table stored in
+  /// \c sourceinfo_block::BasicDeclLocsLayout format.
```
This comment and the one below are no longer accurate.

jrose-apple replied:
Also, please keep all the stored properties together at the top of the class.

jrose-apple commented on lib/Frontend/Frontend.cpp:
```diff
@@ -509,17 +506,34 @@ CompilerInstance::getInputBufferAndModuleDocBufferIfPresent(
+  StringRef fileName = llvm::sys::path::filename(NonPrivatePath);
+  llvm::sys::path::remove_filename(moduleSourceInfoFilePath);
+  llvm::sys::path::append(moduleSourceInfoFilePath, "Project");
+  llvm::sys::path::append(moduleSourceInfoFilePath, fileName);
```
I don't think we need to do the "Project" thing for the partial modules.

jrose-apple commented on lib/Serialization/SourceInfoFormat.h:
```diff
@@ -0,0 +1,94 @@
+using llvm::BCVBR;
+
+/// Magic number for serialized source info files.
+const unsigned char SWIFTSOURCEINFO_SIGNATURE[] = { 0xD6, 0x9C, 0xB7, 0x23 };
```
Where does this number come from? :-)

nkcsgexi replied:
I made these number up. Is there a recommended way?

harlanhaskins replied:
Concept: use the UTF8 bytes of 👩‍⚕️

harlanhaskins replied:
Concept: use the UTF8 bytes of 📒

nkcsgexi replied:
I end up using 🏎️as the magic number :)

jrose-apple commented on lib/Serialization/SourceInfoFormat.h:
```diff
@@ -0,0 +1,94 @@
+///
+/// Increment this value when making a backwards-incompatible change, i.e. where
+/// an \e old compiler will \e not be able to read the new format. This should
+/// be rare. When incrementing this value, reset SWIFTDOC_VERSION_MINOR to 0.
```
Don't forget to update the comment!

jrose-apple commented on lib/Serialization/SourceInfoFormat.h:
```diff
@@ -0,0 +1,94 @@
+
+  using BasicDeclLocsLayout = BCRecordLayout<
+    BASIC_DECL_LOCS, // record ID
+    BCBlob           // an array of fixed size location data
```
Not very extensible, but probably good enough.

jrose-apple commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -2202,34 +2355,63 @@ Optional<CommentInfo> ModuleFile::getCommentForDecl(const Decl *D) const {
+  uint32_t LineColumnCount = 3;
+  uint32_t RecordSize = NumSize + NumSize * 2 * LineColumnCount;
+  uint32_t RecordOffset = RecordSize * UsrId;
+  assert(RecordOffset < BasicDeclLocsData.size());
```
If you want to be extra future-proof, check if `BasicDeclLocsData` is empty first. That would indicate a future version of the swiftsourceinfo file format that doesn't provide BasicDeclLocsData anymore. (Ditto for the other lookups.)

jrose-apple reviewed (commented):
More comments on everything except SerializeDoc.cpp.

nkcsgexi committed:
SerializeLoc: address more comments from Jordan. NFC

jrose-apple commented on lib/Serialization/SerializeDoc.cpp:
```diff
@@ -502,7 +510,300 @@ void serialization::writeDocToStream(raw_ostream &os, ModuleOrSourceFile DC,
+    // We shouldn't expose any Decls that .swiftdoc file isn't willing to expose.
+    // .swiftdoc doesn't include comments for double underscored symbols, but for .swiftsourceinfo,
+    // having the source location for these symbols isn't a concern becuase these
+    // symbols are in .swiftinterface anyway.
```
These two sentences are a little contradictory as written.

jrose-apple reviewed (commented):
Comments on the serialization side, mostly minor.

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented on lib/Serialization/SerializeDoc.cpp:
```diff
@@ -502,7 +510,300 @@ void serialization::writeDocToStream(raw_ostream &os, ModuleOrSourceFile DC,
+
+  Optional<DeclLocationsTableData> getLocData(Decl *D) {
+    auto *File = D->getDeclContext()->getModuleScopeContext();
+    if (auto *SF = dyn_cast<SourceFile>(File)) {
```
Moving this logic to `SourceFile:: getBasicLocsForDecl` raises a question of how to manage the memory of the source file paths in the ASTContext, which probably doesn't worth it.

jrose-apple replied:
`SourceManager::getDisplayNameForLoc` is already returning a StringRef, so we should be okay.

nkcsgexi replied:
We need to call `llvm::sys::fs::make_absolute` on the path, so the StringRef isn't sufficient.

jrose-apple replied:
Hmm. Can we do that at the point of serialization instead of when returning a value? It's still not great that we do it at all.

nkcsgexi replied:
Ok, the most recent commit moved this logic to `SourceFile::getBasicLocsForDecl()`.

nkcsgexi committed:
ASTMangler: EnumElementDecl could also be a DeclContext with a parameter list

nkcsgexi committed:
SerializeLoc: implement SourceFile::getBasicLocsForDecl()

After having this function, the serialization logic doesn't have to
know where the source locations are coming from (.swift or .swiftmodule).

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci Please smoke test compiler performance

nkcsgexi commented:
@swift-ci please test

nkcsgexi commented:
@swift-ci please test

nkcsgexi commented:
Talked with @jrose-apple about this, we plan to merge it for now to unblock progress. Future code review comments, if any, will be addressed by separate PRs.

compnerd commented:
I think that this may have caused a regression on Windows: https://github.com/apple/swift-corelibs-foundation/pull/2521/files

jrose-apple commented on lib/Serialization/SerializeDoc.cpp:
```diff
@@ -502,7 +511,288 @@ void serialization::writeDocToStream(raw_ostream &os, ModuleOrSourceFile DC,
+      Buffer.append(Text);
+      Buffer.push_back('\0');
+    }
+    return IndexMap[Text];
```
Performance nitpick: this will do 2-3 hash lookups of the text, and while that's not *super* important, you can do better with something like this:

```cxx
auto iterAndIsNew = IndexMap.insert({Text, Buffer.size()});
if (iterAndIsNew.second) {
  Buffer.append(Text);
  Buffer.push_back('\0');
}
return (*IterAndIsNew.first).getValue();
```

jrose-apple commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -1222,13 +1223,166 @@ bool ModuleFile::readModuleDocIfPresent() {
+std::unique_ptr<ModuleFile::SerializedDeclUSRTable>
+ModuleFile::readDeclUSRsTable(ArrayRef<uint64_t> fields, StringRef blobData) {
+  if (fields.empty() || blobData.empty())
+     return nullptr;
```
Funky indentation here.

jrose-apple reviewed (commented):
Further comments, none of which are too serious.
